Src/FluentAssertions/Collections/AsyncEnumerableAssertions.cs

```diff
@@ -0,0 +1,3651 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using System.Linq.Expressions;
+using FluentAssertions.Collections.MaximumMatching;
+using FluentAssertions.Common;
+using FluentAssertions.Equivalency;
+using FluentAssertions.Execution;
+using FluentAssertions.Formatting;
+using FluentAssertions.Primitives;
+
+namespace FluentAssertions.Collections;
+
+[DebuggerNonUserCode]
+public class AsyncEnumerableAssertions<T> : AsyncEnumerableAssertions<IAsyncEnumerable<T>, T, AsyncEnumerableAssertions<T>>
```

Comment:
Wouldn't it be much simpler to just have the new `Should` overload enumerate the async collection and then return the existing `GenericCollectionAssertions`?

Reply:
lol yeah you're right. That also allows us to avoid taking dependency on `System.Linq.Async`!

Reply:
ok that actually failed when we have a class that implements multiple `IAsyncEnumerable` generic types. With `RespectingRuntimeTypes` option, it no longer sees the type has having multiple generic types for the `IAsyncEnumerable` interface.

i.e. this fails.

```
    [Fact]
    public void
        When_a_object_implements_multiple_IAsyncEnumerable_interfaces_but_the_declared_type_is_assignable_to_only_one_and_runtime_checking_is_configured_it_should_fail()
    {
        // Arrange
        IAsyncEnumerable<string> collection1 = new AsyncEnumerableOfStringAndObject();
        IAsyncEnumerable<string> collection2 = new AsyncEnumerableOfStringAndObject();

        // Act
        Action act =
            () => collection1.Should().BeEquivalentTo(collection2, opts => opts.RespectingRuntimeTypes());

        // Assert
        act.Should().Throw<XunitException>("the runtime type is assignable to two IAsyncEnumerable interfaces")
            .WithMessage("*cannot determine which one*");
    }
```

Reply:
`AsyncEnumerableAssertions` is gross as it violates DRY having lots of shared code with `GenericCollectionAssertions`. However, I can't see any other way due to the use of inheritance (violating composition over inheritance). :(

Reply:
Ok I've updated `AsyncEnumerableAssertions` to reuse `GenericCollectionAssertions` whereever possible.

Reply:
I've had to revert the above. We lose `ReferenceEquals` checks when we do that as the references are no longer the original one so they are no longer "equal".